### PR TITLE
VC-6776 sheduler and fix payment

### PR DESCRIPTION
**Pull request by alex-bodnar**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: database/block.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// GetBlock - get block from database
+func (db *Db) GetBlock(filter filter.Filter) (types.BlockRow, error) {
+	query, args := filter.SetLimit(1).Build("block")
+
+	var result types.BlockRow
+	if err := db.Sqlx.Get(&result, query, args...); err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return types.BlockRow{}, err
```
Comment (IamStubborN) [inline database/block.go]: errs.Internal?

File: database/block.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	return result, nil
+}
+
+// GetBlocks - get blocks from database
+func (db *Db) GetBlocks(filter filter.Filter) ([]types.BlockRow, error) {
+	query, args := filter.Build("block")
+
+	var val []types.BlockRow
+	if err := db.Sqlx.Select(&val, query, args...); err != nil {
+		return []types.BlockRow{}, err
```
Comment (IamStubborN) [inline database/block.go]: errs.Internal?

File: database/transaction.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+// GetTransaction - get transaction from database
+func (db *Db) GetTransaction(filter filter.Filter) (*txtypes.Tx, error) {
+	query, args := filter.SetLimit(1).Build("transaction")
+
+	var result types.TransactionRow
+	if err := db.Sqlx.Get(&result, query, args...); err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return &txtypes.Tx{}, err
```
Comment (IamStubborN) [inline database/transaction.go]: errs.Internal?

File: database/transaction.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	return db.toTxTypesTx(result)
+}
+
+// GetTransactions - get transactions from database
+func (db *Db) GetTransactions(filter filter.Filter) ([]*txtypes.Tx, error) {
+	query, args := filter.Build("transaction")
+
+	var result []types.TransactionRow
+	if err := db.Sqlx.Select(&result, query, args...); err != nil {
+		return []*txtypes.Tx{}, err
```
Comment (IamStubborN) [inline database/transaction.go]: errs.Internal?

File: database/transaction.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+	if len(result) == 0 {
+		return []*txtypes.Tx{}, errs.NotFound{What: "transaction"}
+	}
+
+	transactions := make([]*txtypes.Tx, 0, len(result))
+	for _, tx := range result {
+		transaction, err := db.toTxTypesTx(tx)
+		if err != nil {
+			return []*txtypes.Tx{}, err
```
Comment (IamStubborN) [inline database/transaction.go]: errs.Internal?

File: database/vipcoin/chain/wallets/set_state.go [hunk 6]
```diff
@@ -75,7 +60,7 @@ func (r Repository) UpdateStates(messages ...*walletstypes.MsgSetWalletState) er
 				 WHERE address = :address`
 
 	for _, m := range messages {
-		stateDB := toSetWalletStateDatabase(m)
+		stateDB := toSetWalletStateDatabase(m, "")
```
Comment (IamStubborN) [inline database/vipcoin/chain/wallets/set_state.go:64]: тут так должно быть?

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 7]
```diff
@@ -31,171 +29,170 @@ func (m *Module) handleMsgPayments(tx *juno.Tx, index int, msg *banking.MsgPayme
...
+		coinFeeRefReward = sdk.NewCoin(asset.Name, sdk.NewIntFromUint64(feeRefReward))
+		coinFeeVoid      = sdk.NewCoin(asset.Name, sdk.NewIntFromUint64(feeVoid))
 	)
 
 	// subtract coins from sender wallet balance
-	walletFrom.Balance, _ = walletFrom.Balance.SafeSub(sdk.NewCoins(coin))
+	walletFrom.Balance = walletFrom.Balance.Sub(sdk.NewCoins(coin))
 	if err := m.walletsRepo.UpdateWallets(&walletFrom); err != nil {
-		return &banking.Payment{}, err
+		return err
```
Comment (IamStubborN) [inline modules/vipcoin/chain/banking/msg_payment.go:156]: errs.Internal? и ниже тоже

**Comment by IamStubborN** [comment]
Лучше везде проставить errs.Internal, тогда по шаблону можно будет в графане эти ошибки находить

**Comment by IamStubborN** [comment]
а так LGTM

File: database/vipcoin/chain/wallets/set_state.go (hunk 6 above)
Comment (alex-bodnar) [reply database/vipcoin/chain/wallets/set_state.go:64]: Да, поле transaction_hash мы не обнавляем, а функция конвертор универсальная

File: database/transaction.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
+package database
```
Comment (Chuiko-GIT) [inline database/transaction.go:1]: тут Copyright не добавляем ?

Comment (alex-bodnar) [reply database/transaction.go:1]: Нет, мы же фркнули bdjuno

File: database/vipcoin/chain/banking/msg_system_reward_transfer.go [hunk 9]
```diff
@@ -7,33 +7,29 @@ import (
...
-// SaveMsgSystemRewardTransfers - method that create transfers to the "vipcoin_chain_banking_system_reward_transfer"
-func (r Repository) SaveMsgSystemRewardTransfers(transfers ...*bankingtypes.MsgSystemRewardTransfer) error {
-	if len(transfers) == 0 {
-		return nil
-	}
-
-	query := `INSERT INTO vipcoin_chain_banking_system_reward_transfer 
-		(creator, wallet_from, wallet_to, asset, amount, extras) 
+// SaveMsgSystemRewardTransfers - method that create transfers to the "vipcoin_chain_banking_system_msg_reward_transfer"
+func (r Repository) SaveMsgSystemRewardTransfers(transfers *bankingtypes.MsgSystemRewardTransfer, transactionHash string) error {
```
Comment (Chuiko-GIT) [inline database/vipcoin/chain/banking/msg_system_reward_transfer.go:12]: можно еще тесты поправить, там идет вызов подобных функций

Comment (alex-bodnar) [reply database/vipcoin/chain/banking/msg_system_reward_transfer.go:12]: готово

**Comment by Chuiko-GIT** [comment]
LGTM

**Pull request merged**